Source/spells.h

```diff
@@ -9,9 +9,17 @@
 
 namespace devilution {
 
+enum class SpellFailReason : uint8_t {
+	Success = 0,
+	None = 0,
+	NoMana,
+	Level0,
+	Busy
+};

```

Comment:
Can't we use a discriminated union instead of an enum to map these?

Also, I really dislike "Fail.Success" type of result (dreaded Windows API ERROR_SUCCESS). If we really need an enum here, I'd recommend something neutral as the enum name, like `SpellCheckingResult`, with a "Success" member (instead of None),

Reply:
what do you mean with `discriminated union`?

Reply:
I know what it is, but not sure how it would apply here.
https://en.wikipedia.org/wiki/Tagged_union

Reply:
same for me - thatswhy I thought that he ment something different

Reply:
Sorry for not replying back. Discriminated unions are done using the `std::variant` type in C++, which signals a value can be one of a closed set of possible results.

Reply:
Are you suggesting to use `HeroSpeech` as the return type?

Reply:
Not really, since the HeroSpeech to use is an implementation detail on the caller and not the responsibility of this method.

Just saying discriminated unions model this kind of scenario better than a raw enum does IMHO. For example, the case of "busy" could also return what the current action is, while the case of "no mana" could return how much mana was needed vs what the player currently has etc.

It is just a more robust way to communicate potentially different return values that is also more extensible.